taichi/program/context.h

```diff
@@ -61,7 +69,7 @@ struct RuntimeContext {
                         DeviceAllocation &alloc,
                         const std::vector<int> &shape) {
     args[arg_id] = taichi_union_cast_with_different_sizes<uint64>(&alloc);
-    set_device_allocation(arg_id, true);
+    set_array_is_device_allocation(arg_id, /*is_device_allocation=*/true);
```

Comment:
We also need to `set_array_runtime_size` here?

Reply:
But why the old code didn't need this?

Reply:
Because we only need the size information to do memcpy for **external arrays** in JIT runtime. This was one of the difference I mentioned between ndarray and external array in runtime that ndarray doesn't need implicit memcpy at this level. Its size information is available at higher `Ndarray` level which is good enough. 
IMHO a better way to do this is creating a class `ExternalArray` which carries its size information and handles the DeviceAllocation alloc/dealloc & memcpy instead of making RuntimeContext more complicated. (That way we can always assume only primitive_types or DeviceAllocation inside RuntimeContext, no need to worry about handling host_ptr inside each backend any more) 
These two functions are primarily used in the AOT runtime where we don't need to worry about external arrays so it was fine :P.

Reply:
> IMHO a better way to do this is creating a class ExternalArray which carries its size information and handles the DeviceAllocation alloc/dealloc & memcpy instead of making RuntimeContext more complicated.

That will be very useful indeed! But my point is that, since the old code doesn't need setting size here yet & this is mostly a refactoring, we probably don't have to `set_array_runtime_size` here?

Reply:
sure, I was mainly afraid of the newly added size info and shape info in extra_args diverge in the `RuntimeContext` :P. But yea it's fine to ignore that for now :D